package/MDAnalysis/analysis/utils/autocorrelation.py

```diff
@@ -0,0 +1,146 @@
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
+#
+# MDAnalysis --- https://www.mdanalysis.org
+# Copyright (c) 2006-2017 The MDAnalysis Development Team and contributors
+# (see the file AUTHORS for the full list of names)
+#
+# Released under the GNU Public Licence, v2 or any higher version
+#
+# Please cite your use of MDAnalysis in published work:
+#
+# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
+# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+# doi: 10.25080/majora-629e541a-00e
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+from __future__ import absolute_import, division
+import numpy as np
+from copy import deepcopy
+
+
+def autocorrelation(list_of_sets, tau_max, window_step=1):
+    r"""The descrete implementation of the autocorrelation function.
+
+    Here is a random equation that shows something.
+
+    .. math::
+       C_{HB}^c(\tau) = \frac{\sum_{ij}h_{ij}(t_0)h'_{ij}(t_0+\tau)}{\sum_{ij}h_{ij}(t_0)}
+
+
+    fixme -  list_of_sets: Modifies in place!
+    Parameters
+    ----------
+    list_of_sets : list
+      List of sets,
+    tau_max : int
+      The last tau (inclusive) for which to carry out autocorrelation.
+    window_step : int, optional
+      The step for the t0 to perform autocorrelation (without the overlap). Default is 1.
+    fixme - move intermittency to the right place
+    intermittency : int, optional
+      Helps with the graps in the data. If we want to remove some of the fluctuations and focus on the patterns,
+      intermittency removes the gaps. The default intermittency=0 which means that if the datapoint is missing at any
+      frame, it is not counted. With the value of 2 the datapoint can be missing for 2 consecutive frames but it will
+      be counted as present.
+
+    Returns
+    --------
+    tau_timeseries : list of int
+        the tau for which the autocorrelation was calculated
+    timeseries : list of int
+        the autocorelation values for each of the tau values
+    timeseries_data : list of list of int
+        the raw data from which the autocorrelation is computed. The time dependant evolution can be investigated.
+
+    .. versionadded:: 0.19.2
+    """
+    # fixme - add checks if this is a list of sets
+    # fixme - check dimensions
+    # fixme - check parameters?
+
+    tau_timeseries = list(range(1, tau_max + 1))
+    timeseries_data = [[] for _ in range(tau_max)]
+
+    # calculate autocorrelation
+    for t in range(0, len(list_of_sets), window_step):
```

Comment:
I'd love to have a look at the cython part since I never wrote any actual code in it. However, it might be important to ask if this is necessary right now. This code has a linear complexity (window_step is always small) and is significantly faster than the previous implementations. The performance tests I run showed that 99.9% of the running time is spent in atom_select, and the actual Autocorrelation took less than a second. 

So I'd advise against making this code more complex with cython, at least until we have a clear case that the performance optimisation here is useful (so on the ground of Premature Optimisation).

Reply:
Hi @richardjgowers , I finally had a look at the performance and I designed this little unit test to see whether I should look more into cython. It took below 1 second to compute 100 k frames with the same atoms (worst case), and around 8 seconds for 1 million frames. With this number of frames, it is going to be atom selection that takes most of the time. For this reason I would recommend against cythonising this function for now. 

Also, it appears that having a list of sets as an argument in the function complicates cythonizing.

```
import time
def test_autocorrelation_perfTest():
    # generate 1 million input, worst case scenario (50 same atoms found each time)
    input = [{x for x in range(50)} for x in range(1000 * 1000)]

    st = time.time()
    autocorrelation(input, tau_max=3)
    print('Comuptation Time', time.time() - st)
```